axon-server-connector/src/main/java/org/axonframework/axonserver/connector/event/LegacyAxonServerEventStorageEngine.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2010-2024. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.axonserver.connector.event;
+
+import com.google.protobuf.ByteString;
+import io.axoniq.axonserver.connector.AxonServerConnection;
+import io.axoniq.axonserver.connector.event.AggregateEventStream;
+import io.axoniq.axonserver.connector.event.AppendEventsTransaction;
+import io.axoniq.axonserver.grpc.MetaDataValue;
+import io.axoniq.axonserver.grpc.SerializedObject;
+import io.axoniq.axonserver.grpc.event.Event;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
+import org.axonframework.common.Context;
+import org.axonframework.common.infra.ComponentDescriptor;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.eventhandling.GenericEventMessage;
+import org.axonframework.eventhandling.GlobalSequenceTrackingToken;
+import org.axonframework.eventhandling.TrackingToken;
+import org.axonframework.eventsourcing.eventstore.AppendCondition;
+import org.axonframework.eventsourcing.eventstore.AppendConditionAssertionException;
+import org.axonframework.eventsourcing.eventstore.AsyncEventStorageEngine;
+import org.axonframework.eventsourcing.eventstore.Index;
+import org.axonframework.eventsourcing.eventstore.IndexedEventMessage;
+import org.axonframework.eventsourcing.eventstore.SourcingCondition;
+import org.axonframework.eventsourcing.eventstore.StreamingCondition;
+import org.axonframework.messaging.MessageStream;
+import org.axonframework.messaging.MetaData;
+import org.axonframework.serialization.Converter;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class LegacyAxonServerEventStorageEngine implements AsyncEventStorageEngine {
+
+    private final AxonServerConnection connection;
+    private final Converter payloadConverter;
+
+    public LegacyAxonServerEventStorageEngine(io.axoniq.axonserver.connector.AxonServerConnection connection,
+                                              Converter payloadConverter) {
+        this.connection = connection;
```

Comment:
Yes, correct. The Event Store will have means to use multi-context. The Storage Engine just uses a single context.
A multi-context Event Store will need to be configured with a Function to create/provide the engine per context.

For the AxonServer Storage Engine it wouldn't be difficult to make it multi-context. But for JPA/JDBC implementation, it would create a lot of overhead to manage the right clients. There, an engine per context would be significantly easier.